### ISSUE TITLE
Hyperparameter Optimization Task Best Practices

### ISSUE THREAD

**Issue by AlanHassen** [pytorch/fairseq#2109]
I was wondering if there are any examples of how to do hyperparameter optimization with fairseq. My idea was to use tune (https://docs.ray.io/en/latest/tune.html) and wire it into a custom task, which calculates the accuracy of an epoch and reports the results to tune. 

I can use the valid_step to evaluate the batch at hand:
```
def valid_step(self, sample, model, criterion): 
    loss, sample_size, logging_output = super().valid_step(sample, model, criterion)
    self._evaluateValidationBatch(self.sequence_generator, sample, model)
    return loss, sample_size, logging_output
```
but I can't evaluate the full sequence after an epoch is finished. It seems, there is no hook for end_epoch? Alternatively, I could put the aggregation into the begin_epoch hook, and always execute +1 epochs, so it gets called at the start of the next epoch:
```
def begin_epoch(self, epoch, model):
    """Hook function called before the start of each epoch."""
    if epoch > 1:
     self.accuracy_scorer.evaluateGeneratedSequence()
    self.accuracy_scorer.calculateSingleScore()
    accuracy = self.accuracy_scorer.result_string()
    logger.info("Accuracy {}: ".format(accuracy))
    reportAccuracyToTune(accuracy)
    self.accuracy_scorer.reset()
```

Is there a better solution for this (i don't want to fork the training loop...) and how do you guys, in general, do hyperparameter optimization? This topic definitely needs some coverage in the documentation.

**Comment by villmow** [comment]
Check  #1828 and #1868 for evaluating accuracy/f1/bleu on generated sequences during validation.

**Comment by dzorlu** [comment]
Hi @AlanHassen - Did you figure it out? I am also working on a problem where I need to do a hp sweep. Seems to me that Hydra needs to be stripped away so that configuration can be passed through Ray making it look like the code [here](https://docs.ray.io/en/master/tune/tutorials/tune-pytorch-cifar.html). 

Is there a smarter way to do hp search with fairseq? Thank you for the awesome library.

**Comment by alexeib** [comment]
have you looked at https://hydra.cc/docs/plugins/ax_sweeper/ or https://hydra.cc/docs/next/plugins/nevergrad_sweeper/ ?

**Comment by MohammedAljahdali** [comment]
Hello @AlanHassen @dzorlu this not related to fairseq, but I am also using hydra in a project and I need to use ray tune, so did you find a way to deal with these two packages together?
